File: typescript/src/paths.ts
/**
 * Filesystem locations and repo-root detection.
 *
 * Mirrors python/src/floo/paths.py - see SPEC.md.
 */
import { execSync } from "node:child_process";
import { mkdirSync } from "node:fs";
import { homedir } from "node:os";
import { resolve, join } from "node:path";

export function stateDir(): string {
  const base = process.env.XDG_STATE_HOME ?? join(homedir(), ".local", "state");
  const d = join(base, "floo");
  mkdirSync(d, { recursive: true });
  return d;
}

export function dbPath(): string {
  return join(stateDir(), "registry.db");
}

export function repoRoot(start: string = process.cwd()): string {
  const cwd = resolve(start);
  try {
    const out = execSync("git rev-parse --show-toplevel", {
      cwd,
      stdio: ["ignore", "pipe", "ignore"],
      encoding: "utf8",
    });
    return resolve(out.trim());
  } catch {
    return cwd;
  }
}

File: src/floo/paths.py
"""Filesystem locations and repo-root detection."""
from __future__ import annotations

import os
import subprocess
from pathlib import Path


def state_dir() -> Path:
    """Return the XDG state directory for floo, creating it if needed."""
    base = os.environ.get("XDG_STATE_HOME") or str(Path.home() / ".local" / "state")
    d = Path(base) / "floo"
    d.mkdir(parents=True, exist_ok=True)
    return d


def db_path() -> Path:
    return state_dir() / "registry.db"


def repo_root(start: Path | None = None) -> Path:
    """Resolve the repo root for the given directory.

    Falls back to the directory itself if not inside a git repo.
    """
    cwd = (start or Path.cwd()).resolve()
    try:
        out = subprocess.run(
            ["git", "rev-parse", "--show-toplevel"],
            cwd=cwd,
            capture_output=True,
            text=True,
            check=True,
        )
        return Path(out.stdout.strip()).resolve()
    except (subprocess.CalledProcessError, FileNotFoundError):
        return cwd
